File: src/main/java/com/example/taskmanager/query/TaskViewProjector.java
package com.example.taskmanager.query;

import com.example.taskmanager.FetchTaskQuery;
import com.example.taskmanager.TaskCompleted;
import com.example.taskmanager.TaskCreated;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.axonframework.eventhandling.EventHandler;
import org.axonframework.queryhandling.QueryHandler;
import org.springframework.stereotype.Component;

@Slf4j
@Component
@RequiredArgsConstructor
public class TaskViewProjector {

  private final TaskViewRepository repository;

  @EventHandler
  public void on(TaskCreated event) {
    TaskView projectedView = new TaskView(event.getId(), event.getType(), event.getCreated(),
        event.getDescription(),
        event.getAssignee(), false);

    log.debug("Projecting {} into {}", event, projectedView);

    repository.save(projectedView);
  }

  @EventHandler
  public void on(TaskCompleted event) {
    log.debug("Projecting {}", event);

    repository.findById(event.getId()).ifPresentOrElse(
        task -> {
          task.setComplete(true);
          repository.save(task);
        },
        () -> {
          throw new IllegalArgumentException("Task not found " + event.getId());
        });
  }

  @QueryHandler
  public TaskView handle(FetchTaskQuery query) {
    return repository.findById(query.getTaskId())
        .orElseThrow(() -> new IllegalArgumentException("Task not found " + query.getTaskId()));
  }
}

File: src/main/java/com/example/taskmanager/query/TaskViewProjector.kt
package com.example.taskmanager.query

import com.example.taskmanager.FetchTaskQuery
import com.example.taskmanager.TaskCompleted
import com.example.taskmanager.TaskCreated
import org.axonframework.eventhandling.EventHandler
import org.axonframework.queryhandling.QueryHandler
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component

@Component
class TaskViewProjector(private val repository: TaskViewRepository) {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    @EventHandler
    fun on(event: TaskCreated) {
        val projectedView = TaskView(
                event.id,
                event.type,
                event.created,
                event.description,
                event.assignee,
                false)

        log.debug("Projecting {} into {}", event, projectedView)

        repository.save(projectedView)
    }

    @EventHandler
    fun on(event: TaskCompleted) {
        log.debug("Projecting {}", event)

        repository.findById(event.id)
                .ifPresentOrElse(
                        { task: TaskView ->
                            task.isComplete = true
                            repository.save(task)
                        }
                )
                { throw IllegalArgumentException("Task not found " + event.id) }
    }

    @QueryHandler
    fun handle(query: FetchTaskQuery): TaskView {
        log.debug("Queried {}", query)

        return repository.findById(query.taskId)
                .orElseThrow { IllegalArgumentException("Task not found " + query.taskId) }
    }
}